tests/pyfunc/test_spark.py

```diff
@@ -1351,19 +1351,22 @@ def test_spark_df_schema_inference_for_map_type(spark):
     expected_schema = Schema(
         [
             ColSpec(Array(DataType.string), "arr"),
-            ColSpec(Object([Property("a", DataType.long), Property("b", DataType.long)]), "map1"),
-            ColSpec(Object([Property("e", Array(DataType.string))]), "map2"),
+            ColSpec(Map(value_type=DataType.long), "map1"),
```

Comment:
This seems spark dataframe containing dicts without defining schema as StructType will always be inferred as MapType, thus `Map` type in the signature. Wondering if we should go with Object type first and then Map type as Object type provides more information 🤔

Reply:
if the spark dataset column dicts only contains several fixed keys, it should use StructType

Reply:
Yes that's true, that's why we need the customer to provide enough data when inferring signature, and that's the case for python dictionaries as well. If you don't provide a key, then we infer that key will not exist, if in the future your data contains that key we should drop it.

Reply:
> If you don't provide a key, then we infer that key will not exist, if in the future your data contains that key we should drop it.

I feel like for this use-case, spark dataframe already designs `StructType` which strictly defined allowed property keys and user should use it.

and spark dataframe `MapType` is designed for use-case that data containing arbitrary and unpredictable keys

Reply:
Do ML models ever operate on a completely arbitrary Map, or do they almost always have some expected set of fields?

Reply:
I checked https://docs.google.com/document/d/1WOAttvOC2D7hkwrOvHXn3YKKcqxNkP4VKzoHmUCTTv8/edit , but I didn't find concrete use-case description for map-type.

Reply:
But I think more of my concern is that:  Spark MapType to Mlflow Object type is an unsafe conversion because MapType data might contains arbitrary keys.

MLflow Schema inferring should be safe (enforce_schema doesn't break with inferred schema), under this premise, Inferring more accurate schema info is the best effort but not the must.

to ensure the accurate schema info, user can explicitly set model signature instead of inferring schema from example data.

Reply:
> But more of my concern is that: Spark MapType to Mlflow Object type conversion is an unsafe conversion because MapType data might contains arbitrary keys.
I agree, but the case is that if you defining a dictionary in a pandas dataframe field and then convert that pandas dataframe to spark dataframe, it always assume it's MapType by default (please correct me if I'm wrong).

> to ensure the accurate schema info, user can explicitly set model signature instead of inferring schema from example data.
Yes exactly, we can not make sure signature inferred from a spark dataframe is 100% correct. Maybe we should add some guidance to ask users to explicitly define the schema either on the spark dataframe or the model, either way should make sure the signature is what they want :)

Reply:
> I agree, but the case is that if you define a dictionary in a pandas dataframe field and then convert that pandas dataframe to spark dataframe, it always assumes it's MapType by default

Yes. Pyspark infers it as MapType for the same safety consideration. We can explicitly setting a column to spark StructType by setting schema config, e.g.
```
dataDictionary = [
        ('James',{'hair':'black','eye':'brown'}),
        ('Michael',{'hair':'brown','eye':None}),
        ('Robert',{'hair':'red','eye':'black'}),
        ('Washington',{'hair':'grey','eye':'grey'}),
        ('Jefferson',{'hair':'brown','eye':''})
        ]
df = spark.createDataFrame(data=dataDictionary, schema = "name string, properties struct<hair: string, eye: string>")
df.printSchema()

df.show()
```

Reply:
@BenWilson2 What do you think of this ? :)

Reply:
I'm of the opinion that the Spark type should be preserved in both directions. If I have a MapType column in spark, that is effectively a generic key-value structure that needs to only adhere to the type constraints of that construct (I have an Array of keys (strings) that maps to values (strings)). The benefit of this structure is its compactness; if I have 95% of my data that is missing an optional key entry, I can avoid the need to define a StructType that populates the resulting data structure with a lot of `null` values for the optional key entry. 

My concern with the mapping of Spark MapType -> MLflow Object type is in the return structure. If I re-convert back to a Spark DF and am expecting a MapType field to be created, but it is now a StructType, this casting would require me to re-convert my StructField back to a MapType field. If the data is signficantly complex (multiple nested StructFields, for instance), then that transformation code would likely require some fairly complex conversion translations to be done in PySpark. 

I feel that we should maintain consistency with Spark Types here. If a user has a MapType column, if the output type of calling inferencing results in a dict structure, when that data is output and submitted back to a Spark DF, the expectation would be to return a MapType column, not a StructType with named fields. 

Can we check with the training team here (who is spearheading the effort for OpenAPI standardization) to ensure that the implementation meets the needs of their efforts?

Reply:
@BenWilson2 I don't fully follow the "re-convert" use case. This implies that the model would have a MapType column *output*, which means we think it's acceptable / common for users to define models that return arbitrary fields. I think this results in a weak / hard-to-interpret API contract for inference.

In general, are there use cases for models that operate on arbitrary inputs and produce arbitrary / varying outputs? Is this a best practice?

If not, I'd propose to validate MapType columns in Spark against Object ColSpec's in MLflow signatures, rather than introducing a Map type to MLflow ColSpec.

Reply:
The scenario that comes to mind is in a classifier or a time series forecasting library. MLflow currently hard-enforces a single output type for these models, while the underlying libraries support a wide array of dict-like (or, even worse, but besides the point, tuple returns) outputs based on inference-time parameters being provided. 
I can return my prediction, but I can also return the probability of class membership for the assigned inferred class, or I can return a map of all classes to their inferred probabilities. 
For time series, we do much the same thing where we restrict the output to only be the range of inferred values, while the underlying library supports varying outputs based on the prediction configuration (I can get 6-sigma style confidence intervals, I can get moving average values, I can get a decomposed trend, etc.). 
I was thinking that if there were such a structure, that could perhaps make those use cases a bit more flexible for users working with custom pyfunc who want to return parameter-based return types. 

I'm not wedded to supporting this functionality; just merely mentioning that there have been many requests for supporting this functionality over the years ("I want my prediction and my probability like I get in scikit-learn"), which was one of the primary reasons for us showing users how to override this behavior with a custom pyfunc tutorial. 

It's far less confusing to established users to not support arbitrary return types, since this is in effect a strong departure from MLflow's stance on static schemas in general for inference. 

@WeichenXu123 let's use the Object mapping to remain consistent with all other input and output types.

Reply:
@BenWilson2 In your example, wouldn’t it be friendlier for the calling application if the model explicitly defined each optional field? Trying to understand if there’s a use case for the model not declaring the specific fields that it could possibly return

Reply:
I think @BenWilson2 want to use map type to make it more flexible, "declaring the specific fields that it could possibly return" might not be convenient in cases @BenWilson2 mentioned above.